Single-pass epoch processing and optimised block processing

michaelsproul opened the pull request:
## Proposed Changes

In preparation for merging `tree-states` down, this PR includes all of the consensus-related changes for epoch and block processing.

The epoch processing changes are covered by a (paper) proof here:

- https://ethresear.ch/t/formally-verified-optimised-epoch-processing/17359

We (@onomatic and I) are in the process of formalising this proof in Isabelle/HOL.

The block processing changes are not covered by a proof, and should be reviewed carefully.

## Notes for reviewers

- `ProgressiveBalancesMode` is deprecated because we use the progressive balances cache everywhere, including for calculating justification and finalization. It is no longer optional.
- This PR is a breaking change because the `previous_epoch_active_gwei` field is removed from the `/lighthouse/validator_inclusion/{epoch}/global` endpoint. It never really should have been in there because justification and finalization only cares about the _current_ epoch total active balance.

## Additional Info

Thanks to @dapplion for extracting this from the main `tree-states` branch, solving `FIXME`s and getting it ready for review :tada:

Main tree-states PR:

- https://github.com/sigp/lighthouse/pull/3206

dapplion committed:
Single-pass epoch processing (#4483, #4573)

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

michaelsproul committed:
Delete unused epoch processing code (#5170)

* Delete unused epoch processing code

* Compare total deltas

* Remove unnecessary apply_pending

* cargo fmt

* Remove newline

michaelsproul committed:
Use epoch cache in block packing (#5223)

michaelsproul committed:
Remove progressive balances mode (#5224)

dapplion committed:
inline inactivity_penalty_quotient_for_state

dapplion committed:
drop previous_epoch_total_active_balance

dapplion committed:
fc lint

dapplion committed:
spec compliant process_sync_aggregate (#15)

* spec compliant process_sync_aggregate

* Update consensus/state_processing/src/per_block_processing/altair/sync_committee.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

---------

Co-authored-by: Michael Sproul <micsproul@gmail.com>

michaelsproul committed:
Delete the participation cache (#16)

dapplion committed:
update help

dapplion committed:
Fix op_pool tests

michaelsproul committed:
Fix fork choice tests

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into epoch-single-pass

michaelsproul committed:
Simplify exit cache (#5280)

michaelsproul committed:
Fix clippy on exit cache

michaelsproul committed:
Clean up single-pass a bit (#5282)

ethDreamer commented on consensus/state_processing/src/per_block_processing/process_operations.rs:
```diff
@@ -145,32 +145,36 @@ pub mod altair_deneb {
+                        validator_participation.add_flag(flag_index)?;
+                        proposer_reward_numerator
+                            .safe_add_assign(state.get_base_reward(index)?.safe_mul(weight)?)?;
+
```
When `validator_slashed == true`, the call to `update_progressive_balances_on_attestation()` is a no-op. It *might* be faster / easier to follow if we remove that argument and just wrap this call in an if statement. Just a minor suggestion, not super attached to it or anything.

michaelsproul replied:
I quite like the logic being contained in the progressive balances cache update function. It forces you to pass `validator_slashed`, which makes you consider it, whereas a manual `if` statement could be accidentally lost or forgotten.

ethDreamer commented on consensus/state_processing/src/per_block_processing/process_operations.rs:
```diff
@@ -127,6 +125,8 @@ pub mod altair_deneb {
         state.build_committee_cache(RelativeEpoch::Current, spec)?;
 
```
We use the `epoch_cache` and `slashings_cache` in this function but I don't see how we ensure they've already been initialized (and initialized to the correct epoch) by this point.

Also, regardless of whether we add lines to ensure the epoch/slashing caches are initialized, should we move all cache initialization checks up one level to `process_attestations()` outside the loop that this function is called within?

michaelsproul replied:
They're initialised by top-level block processing here:

https://github.com/sigp/lighthouse/blob/efb457bfb4aa96e16682ffbb98ec2c87d6e9b2c2/consensus/state_processing/src/per_block_processing.rs#L122-L125

Although it wouldn't hurt to check the initialisation again at the start of attestation processing. I'll make that change

ethDreamer commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1535,7 +1689,7 @@ impl<T: EthSpec> BeaconState<T> {
         }
 
```
Should probably fix the comment at the top of this function (builds committee cache, not epoch cache)

ethDreamer commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1321,7 +1428,7 @@ impl<T: EthSpec> BeaconState<T> {
+        Ok(spec.compute_activation_exit_epoch(epoch)?)
     }
 
     /// Return the churn limit for the current epoch (number of validators who can leave per epoch).
```
This function requires the committee caches are built, not the `epoch_cache` correct?

ethDreamer replied:
The line below should be corrected to say this function uses the *committee* cache rather than saying the `epoch_cache`. Similarly for the comment on the function below.

ethDreamer commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1477,6 +1614,7 @@ impl<T: EthSpec> BeaconState<T> {
         self.update_pubkey_cache()?;
         self.build_exit_cache(spec)?;
+        self.build_slashings_cache()?;
 
```
just making sure we've intentionally omitted building the `epoch_cache` here?

michaelsproul replied:
Yeah unfortunately because the `epoch_cache` requires `state_processing` stuff to build, we can't build it here in the `types` crate

This is why the `AllCaches` trait exists

ethDreamer commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -0,0 +1,628 @@
+    let is_in_inactivity_leak = state.is_in_inactivity_leak(previous_epoch, spec)?;
+    let total_active_balance = state.get_total_active_balance()?;
+    let churn_limit = state.get_churn_limit(spec)?;
+    let activation_churn_limit = state.get_activation_churn_limit(spec)?;
```
The functions `get_total_active_balance()`, `get_churn_limit()`, and `get_activation_churn_limit()` all require the **committee** caches to be built but I don't see anything that ensures this?

ethDreamer replied:
Ah I see they are built in `process_epoch` but not necessarily in all the `_slow()` functions which call `process_epoch_single_pass`

michaelsproul replied:
good catch! addressed in https://github.com/sigp/lighthouse/pull/5386

ethDreamer commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -0,0 +1,628 @@
+        spec,
+    )?;
+
+    if delta.rewards != 0 || delta.penalties != 0 {
```
Why is this if statement here? Does it make this function faster?

michaelsproul replied:
It doesn't help on `stable`, but with `tree-states` we use a copy-on-write reference to the balance, and avoiding the mutation when it isn't necessary is a decent saving. See:

https://github.com/sigp/lighthouse/blob/928915c71885a2439f8fa3b3a22ef6b39e978339/consensus/state_processing/src/per_epoch_processing/single_pass.rs#L355-L359

That `make_mut` call is potentially expensive (it does a `VecMap` insert).

ethDreamer commented on consensus/types/src/chain_spec.rs:
```diff
@@ -496,6 +495,16 @@ impl ChainSpec {
         Hash256::from(domain)
     }
 
+    pub fn compute_activation_exit_epoch(&self, epoch: Epoch) -> Result<Epoch, ArithError> {
```
probably worth putting a comment above this that links to the URL of the spec function

ethDreamer commented on beacon_node/beacon_chain/src/attestation_rewards.rs:
```diff
@@ -134,10 +133,8 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-        process_justification_and_finalization(&state, &participation_cache)?
-            .apply_changes_to_state(&mut state);
-        process_inactivity_updates(&mut state, &participation_cache, spec)?;
+        process_justification_and_finalization(&state)?.apply_changes_to_state(&mut state);
```
`process_justification_and_finalization()` requires the `progressive_balances_cache` to be initialized but from what I can see, the cache is likely uninitialized since the state provided to this function is an arbitrary canonical state freshly loaded from the database.

michaelsproul replied:
good catch, added some more initialisation in https://github.com/sigp/lighthouse/pull/5386

ethDreamer commented on beacon_node/beacon_chain/src/attestation_rewards.rs:
```diff
@@ -134,10 +133,8 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-            .apply_changes_to_state(&mut state);
-        process_inactivity_updates(&mut state, &participation_cache, spec)?;
+        process_justification_and_finalization(&state)?.apply_changes_to_state(&mut state);
+        process_inactivity_updates_slow(&mut state, spec)?;
```
Two potential issues here:
1) `process_inactivity_updates_slow()` calls `process_epoch_single_pass()` which requires the committee caches be built first.
2) The comment on `process_inactivity_updates_slow()` indicates that it's supposed to only be used in testing. By utilizing it here, we are relying on it to provide the data for the `/eth/v1/beacon/rewards/attestations/{epoch}` `beacon-api` endpoint

michaelsproul replied:
1. added calls to build the committee caches in `process_epoch_single_pass`
2. clarified the comment that we're OK to use `_slow`, just not in mainline block/epoch processing

ethDreamer commented on beacon_node/beacon_chain/src/attestation_rewards.rs:
```diff
@@ -147,18 +144,14 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-                    .total_balance()
-                    .map_err(|_| BeaconChainError::AttestationRewardsError)?;
+            let unslashed_participating_balance = state
+                .progressive_balances_cache()
```
this should be okay for now since the `progressive_balances_cache` is initialized by the call to `process_inactivity_updates_slow()` above.. but if you change that due to the concerns in the comment above then you'll need to consider if the `progressive_balances_cache` is initialized.

michaelsproul replied:
added explicit initialisation & clarified that using `_slow` is OK here: https://github.com/sigp/lighthouse/pull/5386

ethDreamer commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1836,6 +1970,10 @@ impl<T: EthSpec> BeaconState<T> {
         Ok(sync_committee)
     }
 
+    pub fn get_base_reward(&self, validator_index: usize) -> Result<u64, EpochCacheError> {
```
May be worth putting a comment that this function will error if the epoch cache isn't initialized (which also implies it will fail if the state is pre-altair).

michaelsproul replied:
Done in https://github.com/sigp/lighthouse/pull/5386

ethDreamer commented on consensus/state_processing/src/common/initiate_validator_exit.rs:
```diff
@@ -8,10 +8,10 @@ pub fn initiate_validator_exit<T: EthSpec>(
+    // whether the validator has already exited, we instead prepare the exit cache and compute the
+    // cheap-to-calculate values from that. *Then* we look up the validator a single time in the
+    // validator tree (expensive), make the check and mutate as appropriate.
 
```
I don't understand how this reordering (building the exit cache first and THEN checking to see if the validator has exited) is an improvement here unless you're just trying to ensure the exit cache gets built for some later use by the calling function?

michaelsproul replied:
It saves a couple of extra lookups of `state.get_validator(index)`, which are becoming more expensive with `tree-states`. I extended the comment in https://github.com/sigp/lighthouse/pull/5386

pawanjay176 commented:
Pushed this to 5.2.0 as we want to get the networking bugfixes release asap.

realbigsean commented on consensus/types/src/epoch_cache.rs:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    pub fn check_validity<E: EthSpec>(
```
The generic here doesn't look used

realbigsean commented on consensus/state_processing/src/epoch_cache.rs:
```diff
@@ -0,0 +1,131 @@
+
+        // State root should already have been filled in by `process_slot`, except in the case
+        // of a `partial_state_advance`.
+        let decision_block_root = latest_block_header.canonical_root();
```
Is it worth adding a check here that the state root isn't `0x00..00`?

realbigsean commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1373,20 +1480,28 @@ impl<T: EthSpec> BeaconState<T> {
-            self.get_effective_balance(*i)
-                .and_then(|bal| Ok(acc.safe_add(bal)?))
-        })?;
+        if epoch != self.current_epoch() && epoch != self.next_epoch()? {
```
From what I can tell `state.current_epoch()` is always being passed in, so I'm not sure why an input `epoch` is necessary. Could you always just use `self.current_epoch()` here? I'm not sure why calculating with `self.next_epoch()` would be allowed here

realbigsean commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1571,40 +1725,15 @@ impl<T: EthSpec> BeaconState<T> {
     /// Note: this function will not build any new committee caches, but will build the total
     /// balance cache if the (new) current epoch cache is initialized.
-    pub fn advance_caches(&mut self, spec: &ChainSpec) -> Result<(), Error> {
+    pub fn advance_caches(&mut self, _spec: &ChainSpec) -> Result<(), Error> {
```
unused arg here

realbigsean commented on consensus/state_processing/src/common/base.rs:
```diff
@@ -1,21 +1,32 @@
-    total_active_balance: u64,
+pub fn get_base_reward(
+    validator_effective_balance: u64,
+    sqrt_total_active_balance: SqrtTotalActiveBalance,
```
could this method also be used to replace `get_base_reward_from_effective_balance`?

michaelsproul replied:
good call, nice simplification and should be faster too!

realbigsean commented on beacon_node/operation_pool/src/lib.rs:
```diff
@@ -252,6 +256,17 @@ impl<T: EthSpec> OperationPool<T> {
     ) -> Result<Vec<Attestation<T>>, OpPoolError> {
+        if let BeaconState::Base(_) = state {
+            // Ok
+        } else {
```
```suggestion
        if !matches(state, BeaconState::Base(_)) {
```

realbigsean commented on beacon_node/operation_pool/src/lib.rs:
```diff
@@ -252,6 +256,17 @@ impl<T: EthSpec> OperationPool<T> {
+        } else {
+            // epoch cache must be initialized to fetch base_reward values in the max_cover score
+            // function. Currently max_cover ignores items on errors. If epoch cache is not
+            // initialized, this function returns Ok([]).
```
```suggestion
            // initialized, this function returns an `OpPoolError`.
```

realbigsean commented on consensus/state_processing/src/per_epoch_processing/epoch_processing_summary.rs:
```diff
@@ -34,10 +100,6 @@ impl<T: EthSpec> EpochProcessingSummary<T> {
             &metrics::PARTICIPATION_PREV_EPOCH_SOURCE_ATTESTING_GWEI_TOTAL,
             self.previous_epoch_source_attesting_balance()? as i64,
         );
```
`PARTICIPATION_PREV_EPOCH_ACTIVE_GWEI_TOTAL` can be deleted unless we want to add this metric back in somewhere

realbigsean reviewed (changes requested):
looking sweet :)

michaelsproul committed:
Address Mark's review of single-pass (#5386)

jimmygchen commented on consensus/types/src/beacon_state.rs:
```diff
@@ -134,7 +147,21 @@ pub enum Error {
         current_epoch: Epoch,
         epoch: Epoch,
     },
+    CommitteeCacheDiffInvalidEpoch {
+        prev_current_epoch: Epoch,
+        current_epoch: Epoch,
+    },
+    CommitteeCacheDiffUninitialized {
+        expected_epoch: Epoch,
+    },
+    DiffAcrossFork {
+        prev_fork: ForkName,
+        current_fork: ForkName,
+    },
+    TotalActiveBalanceDiffUninitialized,
+    MissingImmutableValidator(usize),
```
Where are these used? I can't find usage on the current branch, i guess they're on tree-states?

michaelsproul replied:
I think they may be leftover and unused on tree-states too! Will delete

jimmygchen commented on consensus/types/src/beacon_state.rs:
```diff
@@ -1373,20 +1476,28 @@ impl<T: EthSpec> BeaconState<T> {
     }
 
-    /// Implementation of `get_total_balance`, matching the spec.
+    /// Build the total active balance cache from scratch.
```
```suggestion
    /// Compute the total active balance from scratch.
```

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into epoch-single-pass

michaelsproul committed:
Address Sean's review comments (#5414)

* Address most of Sean's review comments

* Simplify total balance cache building

michaelsproul committed:
Clean up unused junk

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into epoch-single-pass

michaelsproul committed:
More self-review

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into epoch-single-pass

michaelsproul commented:
I've addressed all review comments, and made some more final tweaks after a self-review pass in 1f90350b5fecc57f4862cf8548d400a5e43a41c1

I think this is ready for a green tick and merge!

dapplion committed:
Merge branch 'unstable' into epoch-single-pass

dapplion commented:
Fixed conflicts, and made sure all new data structures use `E: EthSpec`

michaelsproul committed:
Fix imports for beta compiler

michaelsproul committed:
Fix tests, probably

realbigsean reviewed (approved):
🚀

realbigsean commented:
@mergifyio queue
